File: src/main/java/org/saiditnet/redreader/reddit/things/RedditPost.java
/*******************************************************************************
 * This file is part of RedReader.
 *
 * RedReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * RedReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with RedReader.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package org.saiditnet.redreader.reddit.things;


import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Nullable;
import org.saiditnet.redreader.jsonwrap.JsonBufferedObject;

public final class RedditPost implements Parcelable, RedditThingWithIdAndType {

	public String id, name;
	public String title, url, author, domain, subreddit, subreddit_id;
	public int num_comments, score, ups, downs, gilded;
	public boolean archived, over_18, hidden, saved, is_self, clicked, stickied;
	public Object edited;
	public Boolean likes, dislikes;
	public Boolean spoiler;

	public long created, created_utc;

	public String selftext, permalink, link_flair_text, author_flair_text;
	public String thumbnail; // an image URL

	public JsonBufferedObject media;
	@Nullable public String rr_internal_dash_url;

	public RedditPost() {}

	@Nullable
	public String getDashUrl() {

		if(rr_internal_dash_url != null) {
			return rr_internal_dash_url;

		} else if(media != null) {
			try {
				rr_internal_dash_url = media.getObject("reddit_video").getString("fallback_url");

			} catch(final Exception e) {
				rr_internal_dash_url = null;
			}

		}

		return rr_internal_dash_url;
	}

	public String getUrl() {

		if(getDashUrl() != null) {
			return rr_internal_dash_url;
		}

		return url;
	}

	// one of the many reasons why the Android API is awful
	private RedditPost(final Parcel in) {
		id = in.readString();
		name = in.readString();
		title = in.readString();
		url = in.readString();
		author = in.readString();
		domain = in.readString();
		subreddit = in.readString();
		subreddit_id = in.readString();
		num_comments = in.readInt();
		score = in.readInt();
		ups = in.readInt();
		downs = in.readInt();
		gilded = in.readInt();
		archived = in.readInt() == 1;
		over_18 = in.readInt() == 1;
		hidden = in.readInt() == 1;
		saved = in.readInt() == 1;
		is_self = in.readInt() == 1;
		clicked = in.readInt() == 1;
		stickied = in.readInt() == 1;

		final long in_edited = in.readLong();
		if(in_edited == -1) {
			edited = false;
		} else {
			edited = in_edited;
		}

		switch(in.readInt()) {
			case -1: likes = false; break;
			case 0: likes = null; break;
			case 1: likes = true; break;
		}

		switch(in.readInt()) {
			case -1: dislikes = false; break;
			case 0: dislikes = null; break;
			case 1: dislikes = true; break;
		}

		created = in.readLong();
		created_utc = in.readLong();
		selftext = in.readString();
		permalink = in.readString();
		link_flair_text = in.readString();
		author_flair_text = in.readString();
		thumbnail = in.readString();

		switch(in.readInt()) {
			case -1:
				spoiler = false;
				break;
			case 0:
				spoiler = null;
				break;
			case 1:
				spoiler = true;
				break;
		}

		rr_internal_dash_url = in.readString();
	}

	public int describeContents() {
		return 0;
	}

	public void writeToParcel(final Parcel parcel, final int flags) {

		parcel.writeString(id);
		parcel.writeString(name);
		parcel.writeString(title);
		parcel.writeString(url);
		parcel.writeString(author);
		parcel.writeString(domain);
		parcel.writeString(subreddit);
		parcel.writeString(subreddit_id);
		parcel.writeInt(num_comments);
		parcel.writeInt(score);
		parcel.writeInt(ups);
		parcel.writeInt(downs);
		parcel.writeInt(gilded);
		parcel.writeInt(archived ? 1 : 0);
		parcel.writeInt(over_18 ? 1 : 0);
		parcel.writeInt(hidden ? 1 : 0);
		parcel.writeInt(saved ? 1 : 0);
		parcel.writeInt(is_self ? 1 : 0);
		parcel.writeInt(clicked ? 1 : 0);
		parcel.writeInt(stickied ? 1 : 0);

		if(edited instanceof Long) {
			parcel.writeLong((Long)edited);
		} else {
			parcel.writeLong(-1);
		}

		if(likes == null) {
			parcel.writeInt(0);
		} else {
			parcel.writeInt(likes ? 1 : -1);
		}

		if(dislikes == null) {
			parcel.writeInt(0);
		} else {
			parcel.writeInt(dislikes ? 1 : -1);
		}

		parcel.writeLong(created);
		parcel.writeLong(created_utc);
		parcel.writeString(selftext);
		parcel.writeString(permalink);
		parcel.writeString(link_flair_text);
		parcel.writeString(author_flair_text);
		parcel.writeString(thumbnail);

		if(spoiler == null) {
			parcel.writeInt(0);
		} else {
			parcel.writeInt(spoiler ? 1 : -1);
		}

		getDashUrl();
		parcel.writeString(rr_internal_dash_url);
	}

	public static final Parcelable.Creator<RedditPost> CREATOR = new Parcelable.Creator<RedditPost>() {
		public RedditPost createFromParcel(final Parcel in) {
			return new RedditPost(in);
		}

		public RedditPost[] newArray(final int size) {
			return new RedditPost[size];
		}
	};

	@Override
	public String getIdAlone() {
		return id;
	}

	@Override
	public String getIdAndType() {
		return name;
	}
}

File: src/main/java/org/quantumbadger/redreader/reddit/kthings/RedditPost.kt
/*******************************************************************************
 * This file is part of RedReader.
 *
 * RedReader is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * RedReader is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with RedReader.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package org.quantumbadger.redreader.reddit.kthings

import android.os.Parcelable
import android.util.Log
import kotlinx.parcelize.Parcelize
import kotlinx.serialization.Serializable
import org.quantumbadger.redreader.image.RedditGalleryAPI
import org.quantumbadger.redreader.reddit.things.RedditThingWithIdAndType

@Suppress("PropertyName")
@Serializable
@Parcelize
data class RedditPost(
	val id: String,
	val name: RedditIdAndType,
	val url: UrlEncodedString? = null,
	val url_overridden_by_dest: UrlEncodedString? = null,
	val title: UrlEncodedString? = null,
	val author: UrlEncodedString? = null,
	val domain: UrlEncodedString? = null,
	val subreddit: UrlEncodedString,
	val num_comments: Int,
	val score: Int,
	val gilded: Int = 0,
	val crosspost_parent: String? = null,
	val upvote_ratio: Double? = null,
	val archived: Boolean = false,
	val over_18: Boolean = false,
	val hidden: Boolean = false,
	val saved: Boolean = false,
	val is_self: Boolean = false,
	val clicked: Boolean = false,
	val stickied: Boolean = false,
	val can_mod_post: Boolean = false,
	val edited: RedditFieldEdited? = null,
	val likes: Boolean? = null,
	val spoiler: Boolean = false,
	val locked: Boolean = false,
	val created_utc: RedditTimestampUTC,

	val selftext: UrlEncodedString? = null,
	val selftext_html: UrlEncodedString? = null,
	val permalink: UrlEncodedString,

	val link_flair_text: UrlEncodedString? = null,
	val author_flair_text: UrlEncodedString? = null,

	val thumbnail: UrlEncodedString? = null, // an image URL

	val media: Media? = null,

	val preview: Preview? = null,
	val is_video: Boolean = false,

	val distinguished: String? = null,
	val suggested_sort: String? = null, // TODO enum type

	@kotlinx.parcelize.IgnoredOnParcel
	val media_metadata: Map<UrlEncodedString, MaybeParseError<RedditMediaMetadata>>? = null,
	@kotlinx.parcelize.IgnoredOnParcel
	val gallery_data: GalleryData? = null,

	val removed_by_category: String? = null

) : RedditThingWithIdAndType, Parcelable {

	init {
		try {
			RedditGalleryAPI.addToCache(this)
		} catch (t: Throwable) {
			Log.e("RedditPost", "Got error when adding post to cache", t)
		}
	}

	@Serializable
	data class GalleryData(
		val items: List<MaybeParseError<GalleryItem>>
	) {
		@Serializable
		@Parcelize
		data class GalleryItem (
			val media_id: UrlEncodedString,
			val caption: UrlEncodedString? = null,
			val outbound_url: UrlEncodedString? = null,
		) : Parcelable
	}

	@Serializable
	@Parcelize
	data class Media(
		val reddit_video: RedditVideo? = null
	) : Parcelable {
		@Serializable
		@Parcelize
		data class RedditVideo (
			val fallback_url: UrlEncodedString? = null
		) : Parcelable
	}

	@Serializable
	@Parcelize
	data class Preview(
		val enabled: Boolean,
		val images: List<Image>? = null,
		val reddit_video_preview: RedditVideoPreview? = null
	) : Parcelable {
		@Serializable
		@Parcelize
		data class ImageDetails(
			val url: UrlEncodedString,
			val width: Int,
			val height: Int
		) : Parcelable

		@Serializable
		sealed class ImageBase {
			abstract val source: ImageDetails?
			abstract val resolutions: List<ImageDetails>?
		}

		@Serializable
		@Parcelize
		data class Image(
			override val source: ImageDetails? = null,
			override val resolutions: List<ImageDetails>? = null,
			val variants: ImageVariants
		) : ImageBase(), Parcelable

		@Serializable
		@Parcelize
		data class ImageVariants(
			val mp4: ImageVariant? = null
		) : Parcelable

		@Serializable
		@Parcelize
		data class ImageVariant(
			override val source: ImageDetails? = null,
			override val resolutions: List<ImageDetails>? = null
		) : ImageBase(), Parcelable

		@Serializable
		@Parcelize
		data class RedditVideoPreview(
			val fallback_url: UrlEncodedString? = null
		) : Parcelable
	}

	override fun getIdAlone(): String {
		return id
	}

	override fun getIdAndType(): RedditIdAndType {
		return name
	}

	fun findUrl(): String? {

		media?.reddit_video?.fallback_url?.decoded?.apply {
			return this
		}

		if (url?.decoded?.contains(".gif") == true) {
			preview?.images?.get(0)?.variants?.mp4?.source?.url?.decoded?.apply {
				return this
			}
		}

		url_overridden_by_dest?.decoded?.apply {
			return this
		}

		return url?.decoded
	}
}
